File: ios/Classes/SocketManagerProxy.m
//
//  SocketManagerProxy.m
//  ti.socketio
//
//  Created by Jan Vennemann on 14.12.18.
//

#import "SocketManagerProxy.h"
#import "SocketIOClientProxy.h"
#import "TiUtils.h"

#import <SocketIO/SocketIO-Swift.h>

@interface SocketManagerProxy ()

@property BOOL autoConnect;
@property double timeout;

@end

@implementation SocketManagerProxy

- (instancetype)initWithContext:(id<TiEvaluator>)context url:(NSURL *)url options:(NSDictionary *)options;
{
  self = [self _initWithPageContext:context];
  if (self == nil) {
    return nil;
  }

  self.autoConnect = [TiUtils boolValue:@"autoConnect" properties:options def:YES];
  self.timeout = [TiUtils doubleValue:@"timeout" properties:options def:20000.f];

  NSDictionary *nativeOptions = [self convertOptions:options];
#if DEBUG
  DebugLog(@"lookup manager with options %@", nativeOptions);
#endif
  self.manager = [[SocketManager alloc] initWithSocketURL:url config:nativeOptions];

  return self;
}

#pragma mark - Public JS methods

- (SocketIOClientProxy *)socket:(id)args
{
  ENSURE_ARG_COUNT(args, 1);
  NSString *path = [TiUtils stringValue:[args objectAtIndex:0]];

  SocketIOClient *socket;
  if ([path isEqualToString:@"/"]) {
    socket = self.manager.defaultSocket;
  } else {
    socket = [self.manager socketForNamespace:path];
  }
  SocketIOClientProxy *socketProxy = [[SocketIOClientProxy alloc] initWithContext:self.pageContext socket:socket manager:self];
  if (self.autoConnect) {
    if (fabs(self.timeout - 0.0f) < DBL_EPSILON) {
      [socket connect];
    } else {
      [socket connectWithTimeoutAfter:self.timeout / 1000.f withHandler:^(void) {
        [socketProxy fireClientEvent:@"connect_timeout" data:nil];
      }];
    }
  }
  return socketProxy;
}

- (void)open:(id)args
{
  [self connect:args];
}

- (void)connect:(id)args
{
  [self.manager connect];
}

- (void)close:(id)args
{
  [self disconnect:args];
}

- (void)disconnect:(id)args
{
  [self.manager disconnect];
}

- (id)toString:(id)args
{
  return [NSString stringWithFormat: @"SocketManager (%p) %@", self, self.manager];
}

#pragma mark - Private methods

- (NSMutableDictionary *)convertOptions:(NSDictionary *)jsOptions
{
  NSMutableDictionary *options = [NSMutableDictionary dictionaryWithDictionary:jsOptions];

  [self options:options removeUnsupported:@"path" silent:false];
  [self options:options rename:@"reconnection" to:@"reconnects"];
  [self options:options rename:@"reconnectionAttempts" to:@"reconnectAttempts"];
  [self options:options rename:@"reconnectionDelay" to:@"reconnectWait"];
  if (options[@"reconnectWait"] != nil) {
    options[@"reconnectWait"] = @([TiUtils doubleValue:@"reconnectWait" properties:options def:1000.f] / 1000);
  }
  [self options:options rename:@"reconnectionDelayMax" to:@"reconnectWaitMax"];
  if (options[@"reconnectWaitMax"] != nil) {
    options[@"reconnectWaitMax"] = @([TiUtils doubleValue:@"reconnectWaitMax" properties:options def:5000.f] / 1000);
  }
  [self options:options removeUnsupported:@"timeout" silent:true];
  [self options:options removeUnsupported:@"autoConnect" silent:true];
  [self options:options removeUnsupported:@"parser" silent:false];

  NSDictionary *queryParams = [options objectForKey:@"query"];
  if (queryParams != nil) {
    [options setObject:queryParams forKey:@"connectParams"];
    [options removeObjectForKey:@"query"];
  }

  return options;
}

- (void)options:(NSMutableDictionary *)options rename:(NSString *)fromName to:(NSString *)toName
{
  id optionValue = [options objectForKey:fromName];
  if (optionValue == nil) {
    return;
  }

  [options removeObjectForKey:fromName];
  [options setObject:optionValue forKey:toName];
}

- (void)options:(NSMutableDictionary *)options removeUnsupported:(NSString *)optionName silent:(BOOL)silent
{
  id optionValue = [options objectForKey:optionName];
  if (optionValue == nil) {
    return;
  }

  if (!silent) {
    NSLog(@"[WARN] The option %@ is not supported on the iOS client and will be ignored.", optionName);
  }
  [options removeObjectForKey:optionName];
}

@end

File: ios/Classes/SocketManagerProxy.swift
//
//  TiSocketioExampleProxy.swift
//  ti.socketio
//
//  Created by Hans Knöchel
//  Copyright (c) 2022 TiDev. All rights reserved.
//

import UIKit
import TitaniumKit
import SocketIO

@objc(SocketManagerProxy)
class SocketManagerProxy: TiProxy {

  var manager: SocketManager!
  var autoConnect: Bool = false
  var timeout: Double = 2000.0
  
  func _init(withPageContext context: TiEvaluator!, and url: URL, options: [String: Any]) -> Self! {
    super._init(withPageContext: context)
    
    autoConnect = TiUtils.boolValue("autoConnect", properties: options, def: false)
    timeout = TiUtils.doubleValue("timeout", properties: options, def: 2000.0)
    
    let nativeOptions = convertOptions(options)
    manager = SocketManager(socketURL: url, config: nativeOptions)
    
    return self
  }
  
  // MARK: Public API's

  @objc(socket:)
  func socket(_ args: [Any]) -> SocketIOClientProxy {
    guard let path = args.first as? String else {
      fatalError("Missing required parameters")
    }

    let socket: SocketIOClient = {
      if (path == "/") {
        return manager.defaultSocket
      } else {
        return manager.socket(forNamespace: path)
      }
    }()
    
    let socketProxy = SocketIOClientProxy()._init(withPageContext: pageContext, socket: socket, manager: self)!
    if autoConnect {
      if fabs(timeout - 0.0) < Double.ulpOfOne {
        socket.connect()
      } else {
        socket.connect(timeoutAfter: timeout / 1000.0) {
          socketProxy.fireEvent("connect_timeout")
        }
      }
    }
    
    return socketProxy
  }
  
  @objc(open:)
  func `open`(args: [Any]?) {
    self.connect(args: args)
  }
  
  @objc(connect:)
  func connect(args: [Any]?) {
    manager.connect()
  }
  
  @objc(close:)
  func close(args: [Any]?) {
    self.close(args: args)
  }
  
  @objc(disconnect:)
  func disconnect(args: [Any]?) {
    manager.disconnect()
  }
  
  @objc(toString:)
  override func toString(_ args: Any!) -> Any! {
    return "SocketManager (\(self)) \(manager!)"
  }

  // MARK: Private API's
  
  private func convertOptions(_ jsOptions: [String: Any]) -> [String: Any] {
    var options = jsOptions
    
    options = self.options(removeUnsupported: options, optionName: "path", silent: false)
    options = self.options(options, rename: "reconnection", toName: "reconnects")
    options = self.options(options, rename: "reconnectionAttempts", toName: "reconnectAttempts")
    options = self.options(options, rename: "reconnectionDelay", toName: "reconnectWait")
    
    if options["reconnectWait"] != nil {
      options["reconnectWait"] = TiUtils.doubleValue("reconnectWait", properties: options, def: 1000.0) / 1000
    }
    
    options = self.options(options, rename: "reconnectionDelayMax", toName: "reconnectionWaitMax")
    
    if options["reconnectWaitMax"] != nil {
      options["reconnectWaitMax"] = TiUtils.doubleValue("reconnectWaitMax", properties: options, def: 5000.0) / 1000
    }
    
    options = self.options(removeUnsupported: options, optionName: "timeout", silent: true)
    options = self.options(removeUnsupported: options, optionName: "autoConnect", silent: true)
    options = self.options(removeUnsupported: options, optionName: "parser", silent: false)
    
    let queryParams = options["query"]
    if queryParams != nil {
      options["connectParams"] = queryParams
      options.removeValue(forKey: "query")
    }

    return options
  }
  
  private func options(_ options: [String: Any], rename fromName: String, toName: String) -> [String: Any] {
    var optionsCopy = options
    let optionValue = optionsCopy[fromName]
    if optionValue == nil {
      return optionsCopy
    }
    
    optionsCopy.removeValue(forKey: fromName)
    optionsCopy[toName] = optionValue
    return optionsCopy
  }

  private func options(removeUnsupported options: [String: Any], optionName: String, silent: Bool) -> [String: Any] {
    var optionsCopy = options
    let optionValue = optionsCopy[optionName]
    if optionValue == nil {
      return optionsCopy
    }
    
    if (!silent) {
      NSLog("[WARN] The option \(optionName) is not supported on the iOS client and will be ignored.");
    }
    optionsCopy.removeValue(forKey: optionName)
    return optionsCopy
  }
}
